src/annotator/anchoring/test/range-test.js

```diff
@@ -0,0 +1,532 @@
+        const range = createSerializedRange();
+        assert.deepEqual(range.toObject(), {
+          start: '/section[1]/p[1]',
+          startOffset: 0,
+          end: '/section[1]/span[1]/p[1]',
+          endOffset: 6,
+        });
+      });
+    });
+
+    describe('#normalize', () => {
+      it('converts a SerializedRange instance to a NormalizedRange instance', () => {
+        const serializedRange = createSerializedRange();
+        const normalizedRange = serializedRange.normalize(container);
+        assert.isTrue(normalizedRange instanceof Range.NormalizedRange);
+        assert.equal(normalizedRange.start.data, 'text 1');
+        assert.equal(normalizedRange.end.data, 'text 3');
+      });
+
+      it('adjusts starting offset to second text node', () => {
+        const serializedRange = createSerializedRange({
+          start: '/section[1]/p[2]',
+          end: '/section[1]/p[2]',
+          startOffset: 7,
+          endOffset: 14,
+        });
+        const normalizedRange = serializedRange.normalize(container);
+        assert.isTrue(normalizedRange instanceof Range.NormalizedRange);
+        assert.equal(normalizedRange.start.data, 'text 2b');
+        assert.equal(normalizedRange.end.data, 'text 2b');
+      });
+
+      context('when offsets are invalid', () => {
+        it("throws an error if it can't find a valid start offset", () => {
+          const serializedRange = createSerializedRange({
+            startOffset: 99,
+          });
+          assert.throws(() => {
+            serializedRange.normalize(container);
+          }, "Couldn't find offset 99 in element /section[1]/p[1]");
+        });
+
+        it("throws an error if it can't find a valid start offset", () => {
+          const serializedRange = createSerializedRange({
+            startOffset: 1,
+            endOffset: 99,
+          });
+          assert.throws(() => {
+            serializedRange.normalize(container);
+          }, "Couldn't find offset 99 in element /section[1]/span[1]/p[1]");
+        });
+      });
+
+      context('nodeFromXPath() does not return a valid node', () => {
+        let fakeNodeFromXPath;
+
+        beforeEach(() => {
+          fakeNodeFromXPath = sinon.stub();
+          $imports.$mock({
+            './xpath-evaluate': {
+              nodeFromXPath: fakeNodeFromXPath,
+            },
+          });
+        });
+
+        afterEach(() => {
+          $imports.$restore();
+        });
+
+        it('throws a range error if nodeFromXPath() throws an error', () => {
+          fakeNodeFromXPath.throws(new Error('error message'));
+          const serializedRange = createSerializedRange();
+          assert.throws(() => {
+            serializedRange.normalize(container);
+          }, 'Error while finding start node: /section[1]/p[1]: Error: error message');
+        });
+
+        it('throws a range error if nodeFromXPath() does not return a node', () => {
+          fakeNodeFromXPath.returns(null);
```

Comment:
When does `nodeFromXPath` return `null` as opposed to throwing an error?

Reply:
perhaps it does not, I just converted the code over that checked for null hence this test, but upon closer inspection, it does not ever return a null so I'll rip that code out.